// @macropygia/vite-plugin-pug-static/dist/build.js
import fs from 'node:fs';
import path from 'node:path';
import { compileFile } from 'pug';
import { outputLog } from './utils.js';
export const vitePluginPugBuild = (settings) => {
    const { options, locals } = settings;
    const pathMap = new Map();
    return {
        name: 'vite-plugin-pug-build',
        enforce: 'pre',
        apply: 'build',
        resolveId(source) {
            const parsedPath = path.parse(source);
            if (parsedPath.ext === '.pug') {
                const pathAsHtml = path.format({
                    dir: parsedPath.dir,
                    name: parsedPath.name,
                    ext: '.html',
                });
                pathMap.set(pathAsHtml, source);
                return pathAsHtml;
            }
            return null;
        },
        load(id) {
            if (path.extname(id) === '.html') {
                if (pathMap.has(id)) {
                    const compiledTemplate = compileFile(pathMap.get(id), options);
                    const html = compiledTemplate(locals);
                    outputLog('info', 'compiled:', path.relative(process.cwd(), pathMap.get(id)));
                    return html;
                }
                return fs.readFileSync(id, 'utf-8');
            }
            return null;
        },
    };
};
